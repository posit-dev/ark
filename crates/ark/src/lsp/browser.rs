//
// browser.rs
//
// Copyright (C) 2022 Posit Software, PBC. All rights reserved.
//
//

use std::process::Command;

use amalthea::events::PositronEvent;
use amalthea::events::ShowHelpEvent;
use anyhow::Result;
use harp::exec::RFunction;
use harp::object::RObject;
use libR_sys::*;

use crate::interface::KERNEL;

pub static mut PORT: u16 = 0;

#[harp::register]
pub unsafe extern "C" fn ps_browse_url(url: SEXP) -> SEXP {
    match ps_browse_url_impl(url) {
        Ok(_) => Rf_ScalarLogical(1),
        Err(error) => {
            log::error!("{}", error);
            Rf_ScalarLogical(0)
        },
    }
}

unsafe fn handle_help_url(url: &str) -> Result<bool> {
    // Check for help URLs
    let port = RFunction::new("tools", "httpdPort").call()?.to::<i32>()?;
    let prefix = format!("http://127.0.0.1:{}/", port);
    if !url.starts_with(&prefix) {
        return Ok(false);
    }

    // Re-direct the help request to our help proxy server.
    let replacement = format!("http://127.0.0.1:{}/", PORT);

    // Fire an event for the front-end.
    let url = url.replace(prefix.as_str(), replacement.as_str());
    let event = PositronEvent::ShowHelp(ShowHelpEvent {
        kind: "url".to_string(),
        content: url,
        focus: true,
    });

    let kernel = KERNEL.as_ref().unwrap().lock().unwrap();
    kernel.send_event(event);
    Ok(true)
}

unsafe fn ps_browse_url_impl(url: SEXP) -> Result<()> {
    // Extract URL.
    let url = RObject::view(url).to::<String>()?;

    // Handle help server requests.
    if handle_help_url(&url)? {
        return Ok(());
    }

    // TODO: What should we do with other URLs? This is used for opening,
    // for example, web applications (e.g. Shiny) and also interactive plots
    // (e.g. htmlwidgets).
    Command::new("open").arg(url).output()?;
    Ok(())
}
