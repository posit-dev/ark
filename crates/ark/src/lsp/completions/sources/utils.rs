//
// utils.rs
//
// Copyright (C) 2023 Posit Software, PBC. All rights reserved.
//
//

use anyhow::Result;
use harp::error::Error;
use harp::eval::RParseEvalOptions;
use harp::exec::RFunction;
use harp::exec::RFunctionExt;
use harp::object::RObject;
use regex::Regex;
use tower_lsp::lsp_types::CompletionItem;
use tree_sitter::Node;
use tree_sitter::Point;

use crate::lsp::completions::completion_item::completion_item_from_data_variable;
use crate::lsp::document_context::DocumentContext;
use crate::lsp::traits::node::NodeExt;
use crate::lsp::traits::point::PointExt;
use crate::lsp::traits::rope::RopeExt;
use crate::treesitter::NodeType;
use crate::treesitter::NodeTypeExt;

pub(super) fn set_sort_text_by_first_appearance(completions: &mut Vec<CompletionItem>) {
    let size = completions.len();

    // Surely there's a better way to figure out what factor of 10 the `size`
    // fits in, but I can't think of it right now
    let mut width = 1;
    let mut value = 10;

    while size >= value {
        value = value * 10;
        width += 1;
    }

    for (i, item) in completions.iter_mut().enumerate() {
        // Start with existing `sort_text` if one exists
        let text = match &item.sort_text {
            Some(sort_text) => sort_text,
            None => &item.label,
        };
        // Append an integer left padded with `0`s
        let prefix = format!("{:0width$}", i, width = width);
        let sort_text = format!("{prefix}-{text}");
        item.sort_text = Some(sort_text);
    }
}

pub(super) fn set_sort_text_by_words_first(completions: &mut Vec<CompletionItem>) {
    // `_` is considered a word character but we typically want those at the end so:
    // - First `^` for "starts with"
    // - Second `^` for "not the \W_"
    // - `\W_` for "non word characters plus `_`"
    // Result is "starts with any word character except `_`"
    let pattern = Regex::new(r"^[^\W_]").unwrap();

    for item in completions {
        // Start with existing `sort_text` if one exists
        let text = match &item.sort_text {
            Some(sort_text) => sort_text,
            None => &item.label,
        };

        if pattern.is_match(text) {
            item.sort_text = Some(format!("1-{text}"));
        } else {
            item.sort_text = Some(format!("2-{text}"));
        }
    }
}

pub(super) fn filter_out_dot_prefixes(
    context: &DocumentContext,
    completions: &mut Vec<CompletionItem>,
) {
    // Remove completions that start with `.` unless the user explicitly requested them
    let user_requested_dot = context
        .document
        .contents
        .node_slice(&context.node)
        .and_then(|x| Ok(x.to_string().starts_with(".")))
        .unwrap_or(false);

    if !user_requested_dot {
        completions.retain(|x| !x.label.starts_with("."));
    }
}

#[derive(PartialEq, Debug)]
pub(super) enum CallNodePositionType {
    Name,
    Value,
    Ambiguous,
    Outside,
    Unknown,
}

pub(super) fn call_node_position_type(node: &Node, point: Point) -> CallNodePositionType {
    match node.node_type() {
        NodeType::Anonymous(kind) if kind == "(" => {
            if point.is_before_or_equal(node.start_position()) {
                // Before the `(`
                return CallNodePositionType::Outside;
            } else {
                // Must be a name position
                return CallNodePositionType::Name;
            }
        },
        NodeType::Anonymous(kind) if kind == ")" => {
            if point.is_after_or_equal(node.end_position()) {
                // After the `)`
                return CallNodePositionType::Outside;
            } else {
                // Let previous leaf determine type (i.e. did the `)`
                // follow a `=` or a `,`?)
                return call_prev_leaf_position_type(&node, false);
            }
        },
        NodeType::Comma => return CallNodePositionType::Name,
        NodeType::Anonymous(kind) if kind == "=" => return CallNodePositionType::Value,
        // Like `fn(arg<tab>)` or `fn(x = 1, arg<tab>)` (which are ambiguous)
        // or `fn(x = arg<tab>)` (which is clearly a `Value`)
        NodeType::Identifier => return call_prev_leaf_position_type(&node, true),
        _ => {
            // Probably a complex node inside `()`. Typically a `Value`
            // unless we are at the very beginning of the node.

            // For things like `vctrs::vec_sort(x = 1, |2)` where you typed
            // the argument value but want to go back and fill in the name.
            if point == node.start_position() {
                return call_prev_leaf_position_type(&node, false);
            }

            return CallNodePositionType::Value;
        },
    }
}

fn call_prev_leaf_position_type(node: &Node, allow_ambiguous: bool) -> CallNodePositionType {
    let Some(previous) = node.prev_leaf() else {
        // We expect a previous leaf to exist anywhere we use this, so if it
        // doesn't exist then we return this marker type that tells us we should
        // probably investigate our heuristics.
        log::warn!(
            "Expected `node` to have a previous leaf. Is `call_node_position_type()` written correctly?"
        );
        return CallNodePositionType::Unknown;
    };

    let after_open_parenthesis_or_comma = if allow_ambiguous {
        // i.e. `fn(arg<tab>)` or `fn(x, arg<tab>)` where it can be
        // ambiguous whether we are on a `Name` or a `Value`.
        CallNodePositionType::Ambiguous
    } else {
        CallNodePositionType::Name
    };

    match previous.node_type() {
        NodeType::Comma => after_open_parenthesis_or_comma,
        NodeType::Anonymous(kind) if kind == "(" => after_open_parenthesis_or_comma,
        NodeType::Anonymous(kind) if kind == "=" => CallNodePositionType::Value,
        _ => CallNodePositionType::Value,
    }
}

pub(super) fn completions_from_evaluated_object_names(
    name: &str,
    enquote: bool,
) -> Result<Option<Vec<CompletionItem>>> {
    log::info!("completions_from_evaluated_object_names({name:?})");

    let options = RParseEvalOptions {
        forbid_function_calls: true,
        ..Default::default()
    };

    // Try to evaluate the object
    let object = harp::parse_eval(name, options);

    // If we get an `UnsafeEvaluationError` here from setting
    // `forbid_function_calls`, we don't even log that one, as that is
    // expected to happen with complex inputs.
    // If we get a `TryCatchError`, that is typically an 'object not found' error resulting
    // from the user typing pseudocode. Log those at info level without a full backtrace.
    let object = match object {
        Ok(object) => object,
        Err(err) => match err {
            Error::UnsafeEvaluationError(_) => return Ok(None),
            Error::TryCatchError { message, .. } => {
                log::info!("Can't evaluate object: {message}");
                return Ok(None);
            },
            _ => {
                log::error!("Can't evaluate object: {err}");
                return Ok(None);
            },
        },
    };

    let completions = if harp::utils::r_is_matrix(object.sexp) {
        // Special case just for 2D arrays
        completions_from_object_colnames(object, name, enquote)?
    } else {
        completions_from_object_names(object, name, enquote)?
    };

    Ok(Some(completions))
}

pub(super) fn completions_from_object_names(
    object: RObject,
    name: &str,
    enquote: bool,
) -> Result<Vec<CompletionItem>> {
    completions_from_object_names_impl(object, name, enquote, "names")
}

pub(super) fn completions_from_object_colnames(
    object: RObject,
    name: &str,
    enquote: bool,
) -> Result<Vec<CompletionItem>> {
    completions_from_object_names_impl(object, name, enquote, "colnames")
}

fn completions_from_object_names_impl(
    object: RObject,
    name: &str,
    enquote: bool,
    function: &str,
) -> Result<Vec<CompletionItem>> {
    log::info!("completions_from_object_names_impl({object:?})");

    let mut completions = vec![];

    unsafe {
        let element_names = RFunction::new("base", function)
            .add(object)
            .call()?
            .to::<Vec<String>>()?;

        for element_name in element_names {
            match completion_item_from_data_variable(&element_name, name, enquote) {
                Ok(item) => completions.push(item),
                Err(err) => log::error!("{err:?}"),
            }
        }
    }

    Ok(completions)
}

#[cfg(test)]
mod tests {
    use harp::eval::parse_eval_global;
    use tree_sitter::Point;

    use crate::lsp::completions::sources::utils::call_node_position_type;
    use crate::lsp::completions::sources::utils::completions_from_evaluated_object_names;
    use crate::lsp::completions::sources::utils::CallNodePositionType;
    use crate::lsp::document_context::DocumentContext;
    use crate::lsp::documents::Document;
    use crate::r_task;
    use crate::treesitter::NodeType;
    use crate::treesitter::NodeTypeExt;

    #[test]
    fn test_call_node_position_type() {
        // Before `(`, but on it
        let point = Point { row: 0, column: 3 };
        let document = Document::new("fn ()", None);
        let context = DocumentContext::new(&document, point, None);
        assert_eq!(
            context.node.node_type(),
            NodeType::Anonymous(String::from("("))
        );
        assert_eq!(
            call_node_position_type(&context.node, context.point),
            CallNodePositionType::Outside
        );

        // After `)`, but on it
        let point = Point { row: 0, column: 4 };
        let document = Document::new("fn()", None);
        let context = DocumentContext::new(&document, point, None);
        assert_eq!(
            context.node.node_type(),
            NodeType::Anonymous(String::from(")"))
        );
        assert_eq!(
            call_node_position_type(&context.node, context.point),
            CallNodePositionType::Outside
        );

        // After `(`, but on it
        let point = Point { row: 0, column: 3 };
        let document = Document::new("fn()", None);
        let context = DocumentContext::new(&document, point, None);
        assert_eq!(
            context.node.node_type(),
            NodeType::Anonymous(String::from("("))
        );
        assert_eq!(
            call_node_position_type(&context.node, context.point),
            CallNodePositionType::Name
        );

        // After `x`
        let point = Point { row: 0, column: 4 };
        let document = Document::new("fn(x)", None);
        let context = DocumentContext::new(&document, point, None);
        assert_eq!(
            call_node_position_type(&context.node, context.point),
            CallNodePositionType::Ambiguous
        );

        // After `x`
        let point = Point { row: 0, column: 7 };
        let document = Document::new("fn(1, x)", None);
        let context = DocumentContext::new(&document, point, None);
        assert_eq!(
            call_node_position_type(&context.node, context.point),
            CallNodePositionType::Ambiguous
        );

        // Directly after `,`
        let point = Point { row: 0, column: 5 };
        let document = Document::new("fn(x, )", None);
        let context = DocumentContext::new(&document, point, None);
        assert_eq!(context.node.node_type(), NodeType::Comma);
        assert_eq!(
            call_node_position_type(&context.node, context.point),
            CallNodePositionType::Name
        );

        // After `,`, but on `)`
        let point = Point { row: 0, column: 6 };
        let document = Document::new("fn(x, )", None);
        let context = DocumentContext::new(&document, point, None);
        assert_eq!(
            context.node.node_type(),
            NodeType::Anonymous(String::from(")"))
        );
        assert_eq!(
            call_node_position_type(&context.node, context.point),
            CallNodePositionType::Name
        );

        // After `=`
        let point = Point { row: 0, column: 6 };
        let document = Document::new("fn(x =)", None);
        let context = DocumentContext::new(&document, point, None);
        assert_eq!(
            context.node.node_type(),
            NodeType::Anonymous(String::from("="))
        );
        assert_eq!(
            call_node_position_type(&context.node, context.point),
            CallNodePositionType::Value
        );

        // In an expression
        let point = Point { row: 0, column: 4 };
        let document = Document::new("fn(1 + 1)", None);
        let context = DocumentContext::new(&document, point, None);
        assert_eq!(context.node.node_type(), NodeType::Float);
        assert_eq!(
            call_node_position_type(&context.node, context.point),
            CallNodePositionType::Value
        );

        let point = Point { row: 0, column: 8 };
        let document = Document::new("fn(1 + 1)", None);
        let context = DocumentContext::new(&document, point, None);
        assert_eq!(context.node.node_type(), NodeType::Float);
        assert_eq!(
            call_node_position_type(&context.node, context.point),
            CallNodePositionType::Value
        );

        // Right before an expression
        // (special case where we still provide argument completions)
        let point = Point { row: 0, column: 6 };
        let document = Document::new("fn(1, 1 + 1)", None);
        let context = DocumentContext::new(&document, point, None);
        assert_eq!(context.node.node_type(), NodeType::Float);
        assert_eq!(
            call_node_position_type(&context.node, context.point),
            CallNodePositionType::Name
        );

        // After an identifier, before the `)`, with whitespace between them,
        // but on the `)`
        let point = Point { row: 0, column: 5 };
        let document = Document::new("fn(x )", None);
        let context = DocumentContext::new(&document, point, None);
        assert_eq!(
            context.node.node_type(),
            NodeType::Anonymous(String::from(")"))
        );
        assert_eq!(
            call_node_position_type(&context.node, context.point),
            CallNodePositionType::Value
        );

        // After an identifier, before the `)`, with whitespace between them,
        // but on the identifier
        let point = Point { row: 0, column: 4 };
        let document = Document::new("fn(x )", None);
        let context = DocumentContext::new(&document, point, None);
        assert!(context.node.is_identifier());
        assert_eq!(
            call_node_position_type(&context.node, context.point),
            CallNodePositionType::Ambiguous
        );
    }

    #[test]
    fn test_completions_from_evaluated_object_names() {
        r_task(|| {
            // Vector with names
            parse_eval_global("x <- 1:2").unwrap();
            parse_eval_global("names(x) <- c('a', 'b')").unwrap();

            let completions = completions_from_evaluated_object_names("x", false)
                .unwrap()
                .unwrap();
            assert_eq!(completions.len(), 2);
            assert_eq!(completions.get(0).unwrap().label, String::from("a"));
            assert_eq!(completions.get(1).unwrap().label, String::from("b"));

            parse_eval_global("remove(x)").unwrap();

            // Data frame
            parse_eval_global("x <- data.frame(a = 1, b = 2, c = 3)").unwrap();

            let completions = completions_from_evaluated_object_names("x", false)
                .unwrap()
                .unwrap();
            assert_eq!(completions.len(), 3);
            assert_eq!(completions.get(0).unwrap().label, String::from("a"));
            assert_eq!(completions.get(1).unwrap().label, String::from("b"));
            assert_eq!(completions.get(2).unwrap().label, String::from("c"));

            parse_eval_global("remove(x)").unwrap();

            // 1D array with names
            parse_eval_global("x <- array(1:2)").unwrap();
            parse_eval_global("names(x) <- c('a', 'b')").unwrap();

            let completions = completions_from_evaluated_object_names("x", false)
                .unwrap()
                .unwrap();
            assert_eq!(completions.len(), 2);
            assert_eq!(completions.get(0).unwrap().label, String::from("a"));
            assert_eq!(completions.get(1).unwrap().label, String::from("b"));

            parse_eval_global("remove(x)").unwrap();

            // Matrix with column names
            parse_eval_global("x <- array(1, dim = c(1, 1))").unwrap();
            parse_eval_global("rownames(x) <- 'a'").unwrap();
            parse_eval_global("colnames(x) <- 'b'").unwrap();

            let completions = completions_from_evaluated_object_names("x", false)
                .unwrap()
                .unwrap();
            assert_eq!(completions.len(), 1);
            assert_eq!(completions.get(0).unwrap().label, String::from("b"));

            parse_eval_global("remove(x)").unwrap();

            // 3D array with column names
            // We currently decide not to return any names here. It is typically quite
            // ambiguous which axis's names you'd want when working with >=3D arrays.
            // But we did find an object, so we return an empty vector.
            parse_eval_global("x <- array(1, dim = c(1, 1, 1))").unwrap();
            parse_eval_global("rownames(x) <- 'a'").unwrap();
            parse_eval_global("colnames(x) <- 'b'").unwrap();

            let completions = completions_from_evaluated_object_names("x", false)
                .unwrap()
                .unwrap();
            assert!(completions.is_empty());

            parse_eval_global("remove(x)").unwrap();
        })
    }
}
