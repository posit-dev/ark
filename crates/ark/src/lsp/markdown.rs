//
// markdown.rs
//
// Copyright (C) 2022 Posit Software, PBC. All rights reserved.
//
//

#![allow(dead_code)]

use ego_tree::NodeRef;
use scraper::node::Text;
use scraper::ElementRef;
use scraper::Node;
use stdext::join;

pub fn md_codeblock(
    language: &str,
    code: &str,
) -> String {
    join!("``` ", language, "\n", code, "\n", "```", "\n")
}

pub fn md_bold(text: &str) -> String {
    join!("**", text, "**")
}

pub fn md_italic(text: &str) -> String {
    join!("_", text, "_")
}

pub fn md_h1(text: &str) -> String {
    join!("# ", text)
}

pub fn md_h2(text: &str) -> String {
    join!("## ", text)
}

pub fn md_h3(text: &str) -> String {
    join!("### ", text)
}

pub fn md_h4(text: &str) -> String {
    join!("#### ", text)
}

pub fn md_h5(text: &str) -> String {
    join!("###### ", text)
}

pub fn md_h6(text: &str) -> String {
    join!("###### ", text)
}

pub fn md_newline() -> String {
    "\n\n".to_string()
}

pub fn elt_text(node: ElementRef) -> String {
    node.text().collect::<String>()
}

pub fn elt_prev(node: ElementRef) -> Option<ElementRef> {
    for sibling in node.prev_siblings() {
        if let Some(elt) = ElementRef::wrap(sibling) {
            return Some(elt);
        }
    }

    None
}

pub fn elt_next(node: ElementRef) -> Option<ElementRef> {
    for sibling in node.next_siblings() {
        if let Some(elt) = ElementRef::wrap(sibling) {
            return Some(elt);
        }
    }

    None
}

pub struct MarkdownConverter<'a> {
    node: NodeRef<'a, Node>,
}

impl<'a> MarkdownConverter<'a> {
    pub fn new(node: NodeRef<'a, Node>) -> Self {
        MarkdownConverter { node }
    }

    pub fn convert(&self) -> String {
        let mut buffer = String::new();
        self.convert_node(self.node, &mut buffer);
        buffer
    }

    fn convert_node(
        &self,
        node: NodeRef<'a, Node>,
        buffer: &mut String,
    ) {
        if node.value().is_element() {
            let element = ElementRef::wrap(node).unwrap();
            self.convert_element(element, buffer);
        } else if node.value().is_text() {
            let text = node.value().as_text().unwrap();
            self.convert_text(text, buffer);
        }
    }

    fn convert_element(
        &self,
        element: ElementRef<'a>,
        buffer: &mut String,
    ) {
        let name = element.value().name();
        match name {
            "code" => {
                buffer.push('`');
                self.convert_children(element, buffer);
                buffer.push('`');
            },

            "h1" | "h2" | "h3" | "h4" | "h5" | "h6" => {
                let count = name.chars().nth(1).unwrap_or('0').to_digit(10).unwrap_or(0);
                buffer.push_str("#".repeat(count as usize).as_str());
                buffer.push(' ');
                self.convert_children(element, buffer);
            },

            "p" => {
                buffer.push('\n');
                self.convert_children(element, buffer);
                buffer.push('\n');
            },

            "tr" => self.convert_row(element, buffer, |cell, buffer| {
                self.convert_node(*cell, buffer);
            }),

            "ol" => {
                for child in element.children() {
                    if child.value().is_element() {
                        let child = ElementRef::wrap(child).unwrap();
                        buffer.push_str("1. ");
                        self.convert_element(child, buffer);
                    }
                }
            },

            "ul" => {
                for child in element.children() {
                    if child.value().is_element() {
                        let child = ElementRef::wrap(child).unwrap();
                        buffer.push_str("- ");
                        self.convert_element(child, buffer);
                    }
                }
            },

            _ => {
                self.convert_children(element, buffer);
            },
        }
    }

    fn convert_children(
        &self,
        node: ElementRef<'a>,
        buffer: &mut String,
    ) {
        for child in node.children() {
            self.convert_node(child, buffer)
        }
    }

    fn convert_text(
        &self,
        text: &Text,
        buffer: &mut String,
    ) {
        buffer.push_str(text.to_string().as_str())
    }

    fn convert_tr(
        &self,
        element: ElementRef<'a>,
        buffer: &mut String,
    ) {
        self.convert_row(element, buffer, |cell, buffer| {
            self.convert_node(*cell, buffer);
        })
    }

    fn convert_row(
        &self,
        element: ElementRef<'a>,
        buffer: &mut String,
        mut callback: impl FnMut(ElementRef<'a>, &mut String),
    ) {
        buffer.push_str("| ");
        for child in element.children() {
            if child.value().is_element() {
                let child = ElementRef::wrap(child).unwrap();
                let mut contents = String::new();
                callback(child, &mut contents);
                contents = contents.replace("\n", " ");
                buffer.push_str(contents.as_str().trim());
                buffer.push_str(" | ");
            }
        }
        buffer.pop();
    }
}
