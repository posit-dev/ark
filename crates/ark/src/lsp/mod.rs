//
// mod.rs
//
// Copyright (C) 2022 Posit Software, PBC. All rights reserved.
//
//

pub mod backend;
pub mod browser;
pub mod comm;
pub mod completions;
pub mod definitions;
pub mod diagnostics;
pub mod documents;
pub mod editor;
pub mod errors;
pub mod events;
pub mod globals;
pub mod handler;
pub mod help;
pub mod help_proxy;
pub mod hover;
pub mod indexer;
pub mod markdown;
pub mod modules;
pub mod references;
pub mod show_message;
pub mod signature_help;
pub mod symbols;
pub mod traits;
pub mod util;
