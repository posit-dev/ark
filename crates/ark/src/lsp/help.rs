//
// help.rs
//
// Copyright (C) 2022 Posit Software, PBC. All rights reserved.
//
//

use harp::exec::RFunction;
use harp::exec::RFunctionExt;
use harp::utils::r_typeof;
use libr::NILSXP;
use regex::Regex;
use scraper::ElementRef;
use scraper::Html;
use scraper::Selector;
use stdext::push;
use stdext::unwrap;
use stdext::unwrap::IntoResult;
use tower_lsp::lsp_types::MarkupContent;
use tower_lsp::lsp_types::MarkupKind;

use crate::lsp::markdown::*;

pub struct RHtmlHelp {
    html: Html,

    /// Is this help page known to be for a function?
    function: bool,
}

pub enum Status {
    Done,
    KeepGoing,
}

impl RHtmlHelp {
    /// SAFETY: Requires access to the R runtime.
    pub fn from_topic(topic: &str, package: Option<&str>) -> anyhow::Result<Option<Self>> {
        Self::get_help(topic, package).map(|html| {
            html.map(|html| Self {
                html,
                function: false,
            })
        })
    }

    /// SAFETY: Requires access to the R runtime.
    pub fn from_function(name: &str, package: Option<&str>) -> anyhow::Result<Option<Self>> {
        Self::get_help(name, package).map(|html| {
            html.and_then(|html| {
                if Self::is_function(&html) {
                    Some(Self {
                        html,
                        function: true,
                    })
                } else {
                    None
                }
            })
        })
    }

    /// SAFETY: Requires access to the R runtime.
    fn get_help(topic: &str, package: Option<&str>) -> anyhow::Result<Option<Html>> {
        // trim off a package prefix if necessary
        let package = package.map(|s| s.replace("package:", ""));

        // get help document
        let contents = RFunction::from(".ps.help.getHtmlHelpContents")
            .param("topic", topic)
            .param("package", package)
            .call();

        let contents = unwrap!(contents, Err(err) => {
            log::error!("{err:?}");
            return Ok(None);
        });

        // check for NULL (implies no help available)
        if r_typeof(*contents) == NILSXP {
            return Ok(None);
        }

        // parse as html
        let contents = String::try_from(&contents)?;
        let html = Html::parse_document(contents.as_str());

        Ok(Some(html))
    }

    /// Is this a help page for a function?
    ///
    /// Uses a heuristic of looking for a `Usage` section to determine if this looks like
    /// function help or not. Can't look for `Arguments`, as some functions don't have
    /// any!
    fn is_function(x: &Html) -> bool {
        // Find all h3 headers in the document
        let selector = Selector::parse("h3").unwrap();
        let mut headers = x.select(&selector);

        // Do any have a usage section?
        headers.any(|header| header.html() == "<h3>Usage</h3>")
    }

    pub fn topic(&self) -> Option<String> {
        // get topic + title; normally available in first table in the document
        let selector = Selector::parse("table").unwrap();
        let preamble = self.html.select(&selector).next()?;

        // try to get the first cell
        let selector = Selector::parse("td").unwrap();
        let cell = preamble.select(&selector).next()?;
        let preamble = elt_text(cell);

        Some(preamble)
    }

    pub fn title(&self) -> Option<String> {
        let selector = Selector::parse("head > title").unwrap();
        let title = self.html.select(&selector).next()?;
        let mut title = elt_text(title);

        // R prepends 'R: ' to the title, so remove it if that exists
        if title.starts_with("R: ") {
            title.replace_range(0..3, "");
        }

        Some(title)
    }

    #[allow(unused)]
    pub fn section(&self, name: &str) -> Option<Vec<ElementRef>> {
        // find all h3 headers in the document
        let selector = Selector::parse("h3").unwrap();
        let mut headers = self.html.select(&selector);

        // search for the header with the matching name
        let needle = format!("<h2>{}</h2>", name);
        let header = headers.find(|elt| elt.inner_html() == needle);

        let header = match header {
            Some(header) => header,
            None => return None,
        };

        // start collecting elements
        let mut elements: Vec<ElementRef> = Vec::new();
        let mut elt = header;

        loop {
            elt = match elt_next(elt) {
                Some(elt) => elt,
                None => break,
            };

            if matches!(elt.value().name(), "h1" | "h2" | "h3") {
                break;
            }

            elements.push(elt);
        }

        Some(elements)
    }

    /// Find and parse the arguments in the HTML help
    ///
    /// The help file has the structure:
    ///
    /// <h3>Arguments</h3>
    ///
    /// <table>
    /// <tr style="vertical-align: top;"><td><code>parameter</code></td>
    /// <td>
    /// Parameter documentation.
    /// </td></tr>
    ///
    /// Note that parameters might be parsed as part of different, multiple tables;
    /// we need to iterate over all tables after the Arguments header.
    ///
    /// SAFETY: Errors if `self.function` is `false`.
    pub fn parameters(
        &self,
        mut callback: impl FnMut(&Vec<&str>, &ElementRef) -> Status,
    ) -> anyhow::Result<()> {
        if !self.function {
            return Err(anyhow::anyhow!(
                "Called `parameters()` on a topic that isn't a function."
            ));
        }

        let selector = Selector::parse("h3").unwrap();
        let mut headers = self.html.select(&selector);
        let header = headers
            .find(|node| node.html() == "<h3>Arguments</h3>")
            .into_result()?;

        let mut elt = header;
        loop {
            // Get the next element.
            elt = unwrap!(elt_next(elt), None => break);

            // If it's a header, time to stop parsing.
            if elt.value().name() == "h3" {
                break;
            }

            // If it's not a table, skip it.
            if elt.value().name() != "table" {
                continue;
            }

            // Get the cells in this table.
            // I really wish R included classes on these table elements...
            let selector = Selector::parse(r#"tr[style="vertical-align: top;"] > td"#).unwrap();
            let mut cells = elt.select(&selector);

            // Start iterating through pairs of cells.
            loop {
                // Get the parameters. Note that multiple parameters might be contained
                // within a single table cell, so we'll need to split that later.
                let lhs = unwrap!(cells.next(), None => { break });
                let names: String = lhs.text().collect();

                // Get the parameters associated with this description.
                let pattern = Regex::new("\\s*,\\s*").unwrap();
                let names = pattern.split(names.as_str()).collect::<Vec<_>>();

                // Get the parameter description.
                let rhs = unwrap!(cells.next(), None => { break });

                // Execute the callback.
                match callback(&names, &rhs) {
                    Status::Done => return Ok(()),
                    Status::KeepGoing => {},
                };
            }

            // If we got here, we managed to find and parse the argument table.
            break;
        }

        Ok(())
    }

    /// Extract content for an individual parameter by name
    ///
    /// SAFETY: Errors if `self.function` is `false`.
    pub fn parameter(&self, name: &str) -> anyhow::Result<Option<MarkupContent>> {
        if !self.function {
            return Err(anyhow::anyhow!(
                "Called `parameter()` on a topic that isn't a function."
            ));
        }

        let mut result = None;

        self.parameters(|params, node| {
            for param in params {
                if *param == name {
                    result = Some(MarkupContent {
                        kind: MarkupKind::Markdown,
                        value: MarkdownConverter::new(**node).convert(),
                    });
                    return Status::Done;
                }
            }

            return Status::KeepGoing;
        })?;

        Ok(result)
    }

    pub fn markdown(&self) -> anyhow::Result<String> {
        let mut markdown = String::new();

        // add topic
        if let Some(topic) = self.topic() {
            push!(markdown, md_italic(&topic), md_newline());
        }

        if let Some(title) = self.title() {
            push!(markdown, md_h2(&title), md_newline(), "------\n");
        }

        // iterate through the different sections in the help file
        for_each_section(&self.html, |header, elements| {
            // add a title
            let header = elt_text(header);
            markdown.push_str(md_h3(header.as_str()).as_str());
            markdown.push_str(md_newline().as_str());

            // add body
            let body = if matches!(header.as_str(), "Usage" | "Examples") {
                let mut buffer = String::new();
                for elt in elements {
                    if elt.value().name() == "hr" {
                        break;
                    }
                    let code = md_codeblock("r", elt_text(elt).as_str());
                    buffer.push_str(code.as_str());
                }
                buffer
            } else if matches!(header.as_str(), "Arguments") {
                // create a buffer for table output
                let mut buffer = String::new();

                // add an empty header
                buffer.push_str("|     |     |\n");
                buffer.push_str("| --- | --- |");

                // generate the markdown table
                for elt in elements {
                    let converter = MarkdownConverter::new(*elt);
                    let table = converter.convert();
                    buffer.push_str(table.as_str());
                }

                buffer
            } else {
                let mut buffer = String::new();
                for elt in elements {
                    let converter = MarkdownConverter::new(*elt);
                    let markdown = converter.convert();
                    buffer.push_str(markdown.as_str());
                }

                buffer
            };

            markdown.push_str(body.as_str());
            markdown.push_str(md_newline().as_str());
        });

        Ok(markdown)
    }
}

fn for_each_section(doc: &Html, mut callback: impl FnMut(ElementRef, Vec<ElementRef>)) {
    // find all h3 headers in the document
    let selector = Selector::parse("h3").unwrap();
    let headers = doc.select(&selector);

    // iterate through them, and pass each (+ the 'body' of the node) to the callback
    for header in headers {
        // collect all the elements following up to the next header
        let mut elements: Vec<ElementRef> = Vec::new();

        // start with the current header
        let mut elt = header;

        // find the next element -- we might need to skip interleaving nodes
        loop {
            // get the next element (if any)
            elt = unwrap!(elt_next(elt), None => { break });

            // if we find a header, assume that's the start of the next section
            if matches!(elt.value().name(), "h1" | "h2" | "h3") {
                break;
            }

            // add it to our list of elements
            elements.push(elt);
        }

        // execute the callback
        callback(header, elements);
    }
}

#[cfg(test)]
mod tests {
    use crate::lsp::help::RHtmlHelp;
    use crate::lsp::help::Status;
    use crate::r_task;

    #[test]
    fn test_help_from_function() {
        r_task(|| {
            let help = RHtmlHelp::from_function("match", None);
            let help = help.unwrap().unwrap();
            assert!(help.function);

            // Not found at all
            let help = RHtmlHelp::from_function("doesnt_exist", None);
            let help = help.unwrap();
            assert!(help.is_none());

            // Found, but not a function!
            let help = RHtmlHelp::from_function("plotmath", None);
            let help = help.unwrap();
            assert!(help.is_none());
            // It is a topic though
            let help = RHtmlHelp::from_topic("plotmath", None);
            let help = help.unwrap();
            assert!(help.is_some());
        });
    }

    #[test]
    fn test_markdown_conversion() {
        r_task(|| {
            let help = RHtmlHelp::from_function("match", None);
            let help = help.unwrap().unwrap();

            let markdown = help.markdown().unwrap();
            markdown.contains("### Usage");
        });
    }

    #[test]
    fn test_parameters_on_non_functions() {
        r_task(|| {
            let help = RHtmlHelp::from_topic("plotmath", None);
            let help = help.unwrap().unwrap();
            // Errors immediately
            insta::assert_snapshot!(help.parameters(|_, _| Status::Done).unwrap_err());
        });
    }

    #[test]
    fn test_parameter_on_non_functions() {
        r_task(|| {
            let help = RHtmlHelp::from_topic("plotmath", None);
            let help = help.unwrap().unwrap();
            // Errors immediately
            insta::assert_snapshot!(help.parameter("foo").unwrap_err());
        });
    }
}
