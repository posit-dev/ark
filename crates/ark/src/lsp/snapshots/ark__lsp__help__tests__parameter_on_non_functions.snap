---
source: crates/ark/src/lsp/help.rs
expression: "help.parameter(\"foo\").unwrap_err()"
---
Called `parameter()` on a topic that isn't a function.
